### PR TITLE
Add network-bytes-out metric support for CLUSTER SLOT-STATS command (#20)

### PR DESCRIPTION
**Pull request by kyle-yh-kim**
#### Update, July 24th 2024
This PR has been cherry-picked into an existing network-bytes-in PR; https://github.com/valkey-io/valkey/pull/720
Both network-bytes-in and out will be tracked under a single PR. I will now close this PR.

---

The metric tracks network egress bytes under per-slot context, by hooking onto COB buffer mutations.

The metric can be viewed by calling the `CLUSTER SLOT-STATS` command, with sample response attached below;

```
127.0.0.1:6379> cluster slot-stats slotsrange 0 0
1) 1) (integer) 0
   2) 1) "key-count"
      2) (integer) 1
      3) "network-bytes-out"
      4) (integer) 175
```

### CODE REVIEW

**Comment by kyle-yh-kim** [comment]
Currently, the PR aggregates `network-bytes-out` from three different sources;
1. User client's outgoing RESP response, as a result of command invocation.
2. Replica client's outgoing replication stream.
3. Outgoing message due to sharded pubsub publishment.

If this feels too confusing, there are two alternatives;

1. Drop the replication stream and sharded pubsub accounting altogether, and strictly track only user client’s and valkey-server’s interaction.
2. Split the `network-bytes-out` metric into three sub-components;
    ```
   127.0.0.1:6379> cluster slot-stats slotsrange 12539 12539
   1) 1) (integer) 12539
      2) 1) "key-count"
         2) (integer) 1
         3) "network-bytes-out-user"
         4) (integer) 10
         5) "network-bytes-out-replication"
         6) (integer) 20
         7) "network-bytes-out-sharded-pubsub"
         8) (integer) 30
    ```

Personally, I would be opposed to dropping the replication stream and sharded pubsub requirements altogether, as the inclusion of all three components better represents the actual egress bytes. Curious to hear the core team’s thoughts on this.

File: src/cluster_slot_stats.c [hunk 1]
```diff
@@ -113,6 +120,71 @@ static void addReplySortedSlotStats(client *c, slotStatForSort slot_stats[], lon
...
+/* Accumulates egress bytes upon sending RESP responses back to user clients. */
+void clusterSlotStatsAddNetworkBytesOutForUserClient(client *c) {
+    if (!canAddNetworkBytesOut(c)) return;
+
+    serverAssert(c->slot >= 0 && c->slot < CLUSTER_SLOTS);
+    server.cluster->slot_stats[c->slot].network_bytes_out += c->net_output_bytes_curr_cmd;
+}
+
+/* Accumulates egress bytes upon sending replication stream. This only applies for primary nodes. */
+void clusterSlotStatsAddNetworkBytesOutForReplication(int len) {
```
Comment (hpatro) [inline src/cluster_slot_stats.c:148]:
```suggestion
void clusterSlotStatsAddNetworkBytesOutForReplication(size_t len) {
```

File: src/cluster_slot_stats.c [hunk 2]
```diff
@@ -113,6 +120,71 @@ static void addReplySortedSlotStats(client *c, slotStatForSort slot_stats[], lon
...
+
+/* Upon SPUBLISH, two egress events are triggerred.
+ * 1) Internal propagation, for clients that are subscribed to the current node.
+ * 2) External propagation, for other nodes within the same shard (could either be a primary or replica).
+ * This function covers the internal propagation component. */
+void clusterSlotStatsAddNetworkBytesOutForShardedPubSubInternalPropagation(client *c, int slot) {
+    /* For a blocked client, c->slot could be pre-filled.
+     * Thus c->slot is backed-up for restoration after aggregation is completed. */
+    int _slot = c->slot;
+    c->slot = slot;
```
Comment (hpatro) [inline src/cluster_slot_stats.c:164]: Not clearly following this trick and seems dangerous. If we do an early exit on the following statement, we don't restore the value in `c->slot`. Is that intentional ?

File: src/networking.c [hunk 3]
```diff
@@ -2833,6 +2838,7 @@ int processCommandAndResetClient(client *c) {
     client *old_client = server.current_client;
     server.current_client = c;
     if (processCommand(c) == C_OK) {
+        clusterSlotStatsAddNetworkBytesOutForUserClient(c);
```
Comment (hpatro) [inline src/networking.c:2841]: This could also go into `afterCommand` ?

File: src/cluster_slot_stats.c (hunk 2 above)
Comment (kyle-yh-kim) [reply src/cluster_slot_stats.c:164]: Good catch, this was not done intentionally. The command should stay idempotent, irregardless of its early return condition.

Update has been reflected on the other PR, which replaces this one; https://github.com/valkey-io/valkey/pull/720

**Comment by kyle-yh-kim** [comment]
This PR has been cherry-picked into an existing network-bytes-in PR; https://github.com/valkey-io/valkey/pull/720
Both network-bytes-in and out will be tracked under a single PR. I will now close this PR.

**Pull request closed without merging**